wkcuber/api/TiffData/TiffMag.py

```diff
@@ -146,30 +264,90 @@ def close(self):
+            -(-max_indices[0] // tile_size[0]) if tile_size else None
+        )  # ceil division
+        x_indices = range(x_first_index, x_last_index) if tile_size else [None]
+
+        y_first_index = (
+            offset[1] // tile_size[1] if tile_size else None
+        )  # floor division
+        y_last_index = (
+            -(-max_indices[1] // tile_size[1]) if tile_size else None
+        )  # ceil division
+        y_indices = range(y_first_index, y_last_index) if tile_size else [None]
+
+        for x in x_indices:
+            for y in y_indices:
+                for z in range(offset[2], offset[2] + shape[2]):
+                    # calculate the offsets and the size for the x and y coordinate
+                    tile_shape = shape[0:2] + shape[3:4]
+                    offset_in_output_data = tuple(
+                        offset[0:2] * np.equal((x, y), (x_first_index, y_first_index))
+                    )
+                    offset_in_input_data = (0, 0)
+
+                    if tile_size:
+                        tile_top_left_corner = np.array((x, y)) * tile_size
+                        tile_bottom_right_corner = tile_top_left_corner + tile_size
+                        shape_top_left_corner = np.maximum(
+                            offset[0:2], tile_top_left_corner
+                        )
+                        shape_bottom_right = np.minimum(
+                            max_indices[0:2], tile_bottom_right_corner
+                        )
+
+                        offset_in_input_data = shape_top_left_corner - offset[0:2]
+                        offset_in_output_data = tuple(
+                            (offset[0:2] - shape_top_left_corner)
+                            * np.equal((x, y), (x_first_index, y_first_index))
+                        )
+                        tile_shape = tuple(
+                            shape_bottom_right - shape_top_left_corner
+                        ) + tuple(shape[3:4])
+
+                    yield tuple(
+                        (
+                            (x, y, z),
+                            tile_shape,
+                            offset_in_output_data,
+                            offset_in_input_data,
+                        )
+                    )
 
     def has_only_one_channel(self) -> bool:
-        return self.num_channels == 1
+        return self.header.num_channels == 1
 
     def assert_correct_data_format(self, data):
-        if not len(data.shape) == 4:
-            raise AttributeError(
-                "The shape of the provided data does not match the expected shape."
-            )
-        if not data.shape[3] == self.num_channels:
-            raise AttributeError(
-                f"The shape of the provided data does not match the expected shape. (Expected {self.num_channels} channels)"
-            )
-        if not np.dtype(data.dtype) == self.dtype:
+        if self.has_only_one_channel():
+            if not len(data.shape) == 3:
```

Comment:
My idea was to keep the behavior very similar to the behavior of wkw. If I create a WKDataset with 1 channel and then I try to write data (with a shape of `(250, 200, 10, 1)`), I get the following exception: `wkw.wkw.WKWException: Data volume must have 1 channels`.

Because of that, I thought that such actions should also be forbidden in the TiffDataset.

Reply:
Ok, I wasn't aware that wkw forbids such shapes for 1-channel data. Then, it's fine to do the same for TiffDataset, I think. However, in the long run, supporting 4D shapes here (and for wkw) would be more generic in my opinion.